docs/architecture/adr-003-subaccounts.md

```diff
@@ -0,0 +1,88 @@
+# ADR 3: Module SubAccounts
+
+## Changelog
+
+## Context
+
+Currently `ModuleAccount`s must be declared upon supply keeper initialization. In addition to this they don't allow for separation of fungible coins within an account.
+
+New structs should be added so a `ModuleAccount` can dynamically add accounts.
+
+## Decision
+
+Add the following structs into `x/supply`.
+
+### Implementation Changes
+
+Introduce two new structs:
+
+* `ModuleMultiAccount`
+* `SubAccount`
+
+`ModuleMultiAccount` maintains an array of `SubAccount` as well as a list of permissions defined upon initialization of supply keeper (permAddrs).
+It has no pubkey and no limit on the number of `SubAccount`s that can be created.
+Its constructor returns a `ModuleMultiAccount` with no `SubAccount`s.
+`MultiModuleAccount` will implement the Account interface.
+
+```go
+// SetCoins will return an error to prevent ModuleMultiAccount address from having a balance.
+// SubAccounts can only be appended to the SubAccount array.
+// Passively tracks the sum of all SubAccount balances.
+type ModuleMultiAccount struct {
+    Subaccs []SubAccount
+    Permissions []string
+    Coins sdk.Coins // passively track all sub account balances
+
+    CreateSubAccount(pubkey, address) int // returns id of subaccount
+    GetSubAccount(id int) SubAccount
+}
+```
+
+To invalidate a `SubAccount` the `ModuleMultiAccount` calls `SetAccountDisabled` for a `SubAccount`
+
+```go
+// Implements the Account interface. Address is the ModuleMultiAccount address with the id appended.
+// Permissions must be a subset of its ModuleMultiAccount permissions.
+// A disabled account can do withdraws, but cannot recieve any coins.
+type SubAccount struct {
+    Address sdk.AccAddress // MultiAccount (parent) address with index appended
+    ID uint // index of subaccount
+    Permissions []string
```

Comment:
Have we decided on clear semantics of how `ModuleMultiAccount` permissions interact with `SubAccount` permissions.

i.e. Do `SubAccounts` inherit their `ModuleMultiAccount` permissions, are the permissions separate?
If so, what is the point of a `ModuleMultiAccount` having separate permissions if it doesn't directly hold any coins?

Write up on permission semantics necessary imo

Reply:
This is mentioned in the comments above `SubAccount`;- ` Permissions must be a subset of its ModuleMultiAccount permissions.`  I can make this line more clear. 

As for ModuleMultiAccount permissions, upon initialization of supply keeper, the possible permissions for a ModuleAccount (using its name for mapping) must be declared. I think it is useful to use these declared permissions to limit SubAccount permissions. I agree it doesn't make sense for `ModuleMultiAccount` to have permissions if it doesn't hold any coins. Is there another way to enforce these permission restrictions without have a Permissions field `ModuleMultiAccount`?

Reply:
Removed permissions from `MultiModuleAccount` and added permissions section. Let me know if its missing anything.